lib/trino-parquet/src/test/java/io/trino/parquet/reader/TestChunkedInputStream.java

```diff
@@ -0,0 +1,108 @@
+import io.airlift.slice.Slice;
+import io.airlift.slice.Slices;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.util.List;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.airlift.slice.Slices.EMPTY_SLICE;
+import static io.trino.memory.context.AggregatedMemoryContext.newSimpleAggregatedMemoryContext;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.testng.Assert.assertEquals;
+
+public class TestChunkedInputStream
+{
+    @Test
+    public void empty()
+            throws IOException
+    {
+        assertThatThrownBy(() -> input(ImmutableList.of())).isInstanceOf(IllegalArgumentException.class);
+    }
+
+    @Test(dataProvider = "chunks")
+    public void testInput(List<byte[]> chunks)
+            throws IOException
+    {
+        ByteArrayOutputStream out = new ByteArrayOutputStream();
+        for (byte[] chunk : chunks) {
+            out.write(chunk);
+        }
+        byte[] expectedBytes = out.toByteArray();
+        byte[] buffer = new byte[expectedBytes.length + 1];
+
+        List<Slice> slices = chunks.stream().map(Slices::wrappedBuffer).collect(toImmutableList());
+        assertEquals(input(slices).readAllBytes(), expectedBytes);
+        assertEquals(input(slices).getSlice(expectedBytes.length).getBytes(), expectedBytes);
+        assertEquals(readAll(input(slices)), expectedBytes);
+
+        assertEquals(input(slices).readNBytes(expectedBytes.length), expectedBytes);
+
+        assertEquals(input(slices).read(buffer, 0, 0), 0);
+        assertEquals(input(slices).getSlice(0), EMPTY_SLICE);
+
+        if (expectedBytes.length > 0) {
+            // read from one chunk only
+            assertEquals(input(slices).read(buffer, 0, 1), 1);
+            assertEquals(buffer[0], expectedBytes[0]);
+        }
+
+        // rad more than total length
+        ChunkedInputStream input = input(slices);
+        int bytesRead = ByteStreams.read(input, buffer, 0, buffer.length);
+        assertEquals(bytesRead, expectedBytes.length > 0 ? expectedBytes.length : -1);
+
+        // read after input is done returns -1
+        assertEquals(input.read(), -1);
+        // getSlice(0) after input is done returns empty slice
+        assertEquals(input.getSlice(0), EMPTY_SLICE);
+        assertThatThrownBy(() -> input.getSlice(1)).isInstanceOf(IllegalArgumentException.class);
+
+        assertEquals(input.read(buffer, 0, 1), -1);
+    }
+
+    @DataProvider
+    public Object[][] chunks()
+    {
+        return new Object[][] {
+                {ImmutableList.of(new byte[] {1, 2, 3})},
+                {ImmutableList.of(new byte[] {1, 2, 3}, new byte[] {1, 2})},
+                {ImmutableList.of(new byte[] {1, 2, 3}, new byte[] {1}, new byte[] {1, 2})},
+        };
+    }
+
+    private static byte[] readAll(ChunkedInputStream input)
```

Comment:
Could we use `ByteStreams.readAll` here ?

Reply:
this uses on purpose single byte `read()` to test it and `ByteStreams.readAll` use the batch `read(byte[], int, int)`